# Telecom Churn Case Study

With 21 predictor variables we need to predict whether a particular customer will switch to another telecom provider or not. In telecom terminology, this is referred to as churning and not churning, respectively.

## Step 1: Importing and Merging Data

In [1]:
# Suppressing Warnings
import warnings
warnings.filterwarnings('ignore')

In [2]:
# Importing Pandas and NumPy
import pandas as pd, numpy as np

In [3]:
# Importing all datasets
churn_data = pd.read_csv("churn_data.csv")
churn_data.head()

customerID  tenure PhoneService        Contract PaperlessBilling  \
0  7590-VHVEG       1           No  Month-to-month              Yes   
1  5575-GNVDE      34          Yes        One year               No   
2  3668-QPYBK       2          Yes  Month-to-month              Yes   
3  7795-CFOCW      45           No        One year               No   
4  9237-HQITU       2          Yes  Month-to-month              Yes   

               PaymentMethod  MonthlyCharges TotalCharges Churn  
0           Electronic check           29.85        29.85    No  
1               Mailed check           56.95       1889.5    No  
2               Mailed check           53.85       108.15   Yes  
3  Bank transfer (automatic)           42.30      1840.75    No  
4           Electronic check           70.70       151.65   Yes

In [4]:
customer_data = pd.read_csv("customer_data.csv")
customer_data.head()

customerID  gender  SeniorCitizen Partner Dependents
0  7590-VHVEG  Female              0     Yes         No
1  5575-GNVDE    Male              0      No         No
2  3668-QPYBK    Male              0      No         No
3  7795-CFOCW    Male              0      No         No
4  9237-HQITU  Female              0      No         No

In [5]:
internet_data = pd.read_csv("internet_data.csv")
internet_data.head()

customerID     MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  7590-VHVEG  No phone service             DSL             No          Yes   
1  5575-GNVDE                No             DSL            Yes           No   
2  3668-QPYBK                No             DSL            Yes          Yes   
3  7795-CFOCW  No phone service             DSL            Yes           No   
4  9237-HQITU                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies  
0               No          No          No              No  
1              Yes          No          No              No  
2               No          No          No              No  
3              Yes         Yes          No              No  
4               No          No          No              No

### Combining all data files into one consolidated dataframe

In [6]:
# Merging on "customerID"
df_1 = pd.merge(churn_data, customer_data, how='inner', on='customerID')

In [7]:
# Final dataframe with all predictor variables
telecom = pd.merge(df_1, internet_data, how='inner', on='customerID')

## Step 2: Inspecting the Dataframe

In [8]:
# Head of our master dataset
telecom.head()

customerID  tenure PhoneService        Contract PaperlessBilling  \
0  7590-VHVEG       1           No  Month-to-month              Yes   
1  5575-GNVDE      34          Yes        One year               No   
2  3668-QPYBK       2          Yes  Month-to-month              Yes   
3  7795-CFOCW      45           No        One year               No   
4  9237-HQITU       2          Yes  Month-to-month              Yes   

               PaymentMethod  MonthlyCharges TotalCharges Churn  gender  ...  \
0           Electronic check           29.85        29.85    No  Female  ...   
1               Mailed check           56.95       1889.5    No    Male  ...   
2               Mailed check           53.85       108.15   Yes    Male  ...   
3  Bank transfer (automatic)           42.30      1840.75    No    Male  ...   
4           Electronic check           70.70       151.65   Yes  Female  ...   

   Partner Dependents     MultipleLines InternetService OnlineSecurity  \
0      Yes         No  No phone service             DSL             No   
1       No         No                No             DSL            Yes   
2       No         No                No             DSL            Yes   
3       No         No  No phone service             DSL            Yes   
4       No         No                No     Fiber optic             No   

  OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  
0          Yes               No          No          No              No  
1           No              Yes          No          No              No  
2          Yes               No          No          No              No  
3           No              Yes         Yes          No              No  
4           No               No          No          No              No  

[5 rows x 21 columns]

In [9]:
# Check the dimensions of the dataframe
telecom.shape

(7043, 21)

In [10]:
# Statistics aspects of the dataframe
telecom.describe()

tenure  MonthlyCharges  SeniorCitizen
count  7043.000000     7043.000000    7043.000000
mean     32.371149       64.761692       0.162147
std      24.559481       30.090047       0.368612
min       0.000000       18.250000       0.000000
25%       9.000000       35.500000       0.000000
50%      29.000000       70.350000       0.000000
75%      55.000000       89.850000       0.000000
max      72.000000      118.750000       1.000000

In [11]:
# Let's see the type of each column
telecom.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   tenure            7043 non-null   int64  
 2   PhoneService      7043 non-null   object 
 3   Contract          7043 non-null   object 
 4   PaperlessBilling  7043 non-null   object 
 5   PaymentMethod     7043 non-null   object 
 6   MonthlyCharges    7043 non-null   float64
 7   TotalCharges      7043 non-null   object 
 8   Churn             7043 non-null   object 
 9   gender            7043 non-null   object 
 10  SeniorCitizen     7043 non-null   int64  
 11  Partner           7043 non-null   object 
 12  Dependents        7043 non-null   object 
 13  MultipleLines     7043 non-null   object 
 14  InternetService   7043 non-null   object 
 15  OnlineSecurity    7043 non-null   object 
 16  OnlineBackup      7043 non-null   object 


## Step 3: Data Preparation

### Converting some binary variables (Yes/No) to 0/1

In [12]:
varlist = ['PhoneService', 'PaperlessBilling', 'Churn', 'Partner', 'Dependents']

# Defining the map function
def binary_map(x):
    return x.map({'Yes':1, 'No':0})

# Applying the function to the housing list
telecom[varlist] = telecom[varlist].apply(binary_map)

In [13]:
telecom.head()

customerID  tenure  PhoneService        Contract  PaperlessBilling  \
0  7590-VHVEG       1             0  Month-to-month                 1   
1  5575-GNVDE      34             1        One year                 0   
2  3668-QPYBK       2             1  Month-to-month                 1   
3  7795-CFOCW      45             0        One year                 0   
4  9237-HQITU       2             1  Month-to-month                 1   

               PaymentMethod  MonthlyCharges TotalCharges  Churn  gender  ...  \
0           Electronic check           29.85        29.85      0  Female  ...   
1               Mailed check           56.95       1889.5      0    Male  ...   
2               Mailed check           53.85       108.15      1    Male  ...   
3  Bank transfer (automatic)           42.30      1840.75      0    Male  ...   
4           Electronic check           70.70       151.65      1  Female  ...   

   Partner  Dependents     MultipleLines InternetService OnlineSecurity  \
0        1           0  No phone service             DSL             No   
1        0           0                No             DSL            Yes   
2        0           0                No             DSL            Yes   
3        0           0  No phone service             DSL            Yes   
4        0           0                No     Fiber optic             No   

  OnlineBackup DeviceProtection TechSupport StreamingTV StreamingMovies  
0          Yes               No          No          No              No  
1           No              Yes          No          No              No  
2          Yes               No          No          No              No  
3           No              Yes         Yes          No              No  
4           No               No          No          No              No  

[5 rows x 21 columns]

### For categorical variables with multiple levels, create dummy features (one-hot encoded)

In [18]:
# Creating a dummy variable for some of the categorical variables and dropping the first one.
dummy1 = pd.get_dummies(telecom[['Contract', 'PaymentMethod', 'gender', 'InternetService']], drop_first=True)

# Adding the results to the master dataframe
telecom = pd.concat([telecom, dummy1], axis=1)

In [19]:
telecom.head()

customerID  tenure  PhoneService        Contract  PaperlessBilling  \
0  7590-VHVEG       1             0  Month-to-month                 1   
1  5575-GNVDE      34             1        One year                 0   
2  3668-QPYBK       2             1  Month-to-month                 1   
3  7795-CFOCW      45             0        One year                 0   
4  9237-HQITU       2             1  Month-to-month                 1   

               PaymentMethod  MonthlyCharges TotalCharges  Churn  gender  ...  \
0           Electronic check           29.85        29.85      0  Female  ...   
1               Mailed check           56.95       1889.5      0    Male  ...   
2               Mailed check           53.85       108.15      1    Male  ...   
3  Bank transfer (automatic)           42.30      1840.75      0    Male  ...   
4           Electronic check           70.70       151.65      1  Female  ...   

   StreamingTV  StreamingMovies  Contract_One year Contract_Two year  \
0           No               No                  0                 0   
1           No               No                  1                 0   
2           No               No                  0                 0   
3           No               No                  1                 0   
4           No               No                  0                 0   

  PaymentMethod_Credit card (automatic) PaymentMethod_Electronic check  \
0                                     0                              1   
1                                     0                              0   
2                                     0                              0   
3                                     0                              0   
4                                     0                              1   

  PaymentMethod_Mailed check gender_Male InternetService_Fiber optic  \
0                          0           0                           0   
1                          1           1                           0   
2                          1           1                           0   
3                          0           1                           0   
4                          0           0                           1   

  InternetService_No  
0                  0  
1                  0  
2                  0  
3                  0  
4                  0  

[5 rows x 29 columns]

In [20]:
# Creating dummy variables for the remaining categorical variables and dropping the level with big names.

# Creating dummy variables for the variable 'MultipleLines'
ml = pd.get_dummies(telecom['MultipleLines'], prefix='MultipleLines')
# Dropping MultipleLines_No phone service column
ml1 = ml.drop(['MultipleLines_No phone service'], 1)
#Adding the results to the master dataframe
telecom = pd.concat([telecom,ml1], axis=1)

# Creating dummy variables for the variable 'OnlineSecurity'.
os = pd.get_dummies(telecom['OnlineSecurity'], prefix='OnlineSecurity')
os1 = os.drop(['OnlineSecurity_No internet service'], 1)
# Adding the results to the master dataframe
telecom = pd.concat([telecom,os1], axis=1)

# Creating dummy variables for the variable 'OnlineBackup'.
ob = pd.get_dummies(telecom['OnlineBackup'], prefix='OnlineBackup')
ob1 = ob.drop(['OnlineBackup_No internet service'], 1)
# Adding the results to the master dataframe
telecom = pd.concat([telecom,ob1], axis=1)

# Creating dummy variables for the variable 'DeviceProtection'. 
dp = pd.get_dummies(telecom['DeviceProtection'], prefix='DeviceProtection')
dp1 = dp.drop(['DeviceProtection_No internet service'], 1)
# Adding the results to the master dataframe
telecom = pd.concat([telecom,dp1], axis=1)

# Creating dummy variables for the variable 'TechSupport'. 
ts = pd.get_dummies(telecom['TechSupport'], prefix='TechSupport')
ts1 = ts.drop(['TechSupport_No internet service'], 1)
# Adding the results to the master dataframe
telecom = pd.concat([telecom,ts1], axis=1)

# Creating dummy variables for the variable 'StreamingTV'.
st =pd.get_dummies(telecom['StreamingTV'], prefix='StreamingTV')
st1 = st.drop(['StreamingTV_No internet service'], 1)
# Adding the results to the master dataframe
telecom = pd.concat([telecom,st1], axis=1)

# Creating dummy variables for the variable 'StreamingMovies'. 
sm = pd.get_dummies(telecom['StreamingMovies'], prefix='StreamingMovies')
sm1 = sm.drop(['StreamingMovies_No internet service'], 1)
# Adding the results to the master dataframe
telecom = pd.concat([telecom,sm1], axis=1)

In [21]:
telecom.head()

customerID  tenure  PhoneService        Contract  PaperlessBilling  \
0  7590-VHVEG       1             0  Month-to-month                 1   
1  5575-GNVDE      34             1        One year                 0   
2  3668-QPYBK       2             1  Month-to-month                 1   
3  7795-CFOCW      45             0        One year                 0   
4  9237-HQITU       2             1  Month-to-month                 1   

               PaymentMethod  MonthlyCharges TotalCharges  Churn  gender  ...  \
0           Electronic check           29.85        29.85      0  Female  ...   
1               Mailed check           56.95       1889.5      0    Male  ...   
2               Mailed check           53.85       108.15      1    Male  ...   
3  Bank transfer (automatic)           42.30      1840.75      0    Male  ...   
4           Electronic check           70.70       151.65      1  Female  ...   

   OnlineBackup_No  OnlineBackup_Yes  DeviceProtection_No  \
0                0                 1                    1   
1                1                 0                    0   
2                0                 1                    1   
3                1                 0                    0   
4                1                 0                    1   

  DeviceProtection_Yes TechSupport_No TechSupport_Yes StreamingTV_No  \
0                    0              1               0              1   
1                    1              1               0              1   
2                    0              1               0              1   
3                    1              0               1              1   
4                    0              1               0              1   

  StreamingTV_Yes StreamingMovies_No StreamingMovies_Yes  
0               0                  1                   0  
1               0                  1                   0  
2               0                  1                   0  
3               0                  1                   0  
4               0                  1                   0  

[5 rows x 43 columns]

### Dropping the repeated variables

In [22]:
# We have created dummies for the below variables, so we can drop them
telecom = telecom.drop(['Contract','PaymentMethod','gender','MultipleLines','InternetService', 'OnlineSecurity', 'OnlineBackup', 'DeviceProtection',
       'TechSupport', 'StreamingTV', 'StreamingMovies'], 1)

In [32]:
#The varaible was imported as a string we need to convert it to float
telecom['TotalCharges'] = pd.to_numeric(telecom.TotalCharges, errors='coerce')

In [33]:
telecom.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7043 entries, 0 to 7042
Data columns (total 32 columns):
 #   Column                                 Non-Null Count  Dtype  
---  ------                                 --------------  -----  
 0   customerID                             7043 non-null   object 
 1   tenure                                 7043 non-null   int64  
 2   PhoneService                           7043 non-null   int64  
 3   PaperlessBilling                       7043 non-null   int64  
 4   MonthlyCharges                         7043 non-null   float64
 5   TotalCharges                           7032 non-null   float64
 6   Churn                                  7043 non-null   int64  
 7   SeniorCitizen                          7043 non-null   int64  
 8   Partner                                7043 non-null   int64  
 9   Dependents                             7043 non-null   int64  
 10  Contract_One year                      7043 non-null   uint8  
 11  Cont

Now you can see that you have all variables as numeric.

### Checking for Outliers

In [34]:
# Checking for outliers in the continuous variables
num_telecom = telecom[['tenure','MonthlyCharges','SeniorCitizen','TotalCharges']]

In [35]:
# Checking outliers at 25%, 50%, 75%, 90%, 95% and 99%
num_telecom.describe(percentiles=[.25, .5, .75, .90, .95, .99])

tenure  MonthlyCharges  SeniorCitizen  TotalCharges
count  7043.000000     7043.000000    7043.000000   7032.000000
mean     32.371149       64.761692       0.162147   2283.300441
std      24.559481       30.090047       0.368612   2266.771362
min       0.000000       18.250000       0.000000     18.800000
25%       9.000000       35.500000       0.000000    401.450000
50%      29.000000       70.350000       0.000000   1397.475000
75%      55.000000       89.850000       0.000000   3794.737500
90%      69.000000      102.600000       1.000000   5976.640000
95%      72.000000      107.400000       1.000000   6923.590000
99%      72.000000      114.729000       1.000000   8039.883000
max      72.000000      118.750000       1.000000   8684.800000

From the distribution shown above, you can see that there are no outliers in data. The numbers are gradually increasing.

### Checking for Missing Values and Inputing Them

In [36]:
# Adding up the missing values (column-wise)
telecom.isnull().sum()

customerID                                0
tenure                                    0
PhoneService                              0
PaperlessBilling                          0
MonthlyCharges                            0
TotalCharges                             11
Churn                                     0
SeniorCitizen                             0
Partner                                   0
Dependents                                0
Contract_One year                         0
Contract_Two year                         0
PaymentMethod_Credit card (automatic)     0
PaymentMethod_Electronic check            0
PaymentMethod_Mailed check                0
gender_Male                               0
InternetService_Fiber optic               0
InternetService_No                        0
MultipleLines_No                          0
MultipleLines_Yes                         0
OnlineSecurity_No                         0
OnlineSecurity_Yes                        0
OnlineBackup_No                 

It means that 11/7043 = 0.001561834 i.e 0.1%, best is to remove these observations from the analysis

In [37]:
# Removing NaN TotalCharges rows
telecom = telecom[~np.isnan(telecom['TotalCharges'])]

In [38]:
# Checking percentage of missing values after removing the missing values
round(100*(telecom.isnull().sum()/len(telecom.index)), 2)

customerID                               0.0
tenure                                   0.0
PhoneService                             0.0
PaperlessBilling                         0.0
MonthlyCharges                           0.0
TotalCharges                             0.0
Churn                                    0.0
SeniorCitizen                            0.0
Partner                                  0.0
Dependents                               0.0
Contract_One year                        0.0
Contract_Two year                        0.0
PaymentMethod_Credit card (automatic)    0.0
PaymentMethod_Electronic check           0.0
PaymentMethod_Mailed check               0.0
gender_Male                              0.0
InternetService_Fiber optic              0.0
InternetService_No                       0.0
MultipleLines_No                         0.0
MultipleLines_Yes                        0.0
OnlineSecurity_No                        0.0
OnlineSecurity_Yes                       0.0
OnlineBack

## Step 4: Test-Train Split

In [39]:
from sklearn.model_selection import train_test_split

In [40]:
# Putting feature variable to X
X = telecom.drop(['Churn', 'customerID'], axis=1)
X.head()

tenure  PhoneService  PaperlessBilling  MonthlyCharges  TotalCharges  \
0       1             0                 1           29.85         29.85   
1      34             1                 0           56.95       1889.50   
2       2             1                 1           53.85        108.15   
3      45             0                 0           42.30       1840.75   
4       2             1                 1           70.70        151.65   

   SeniorCitizen  Partner  Dependents  Contract_One year  Contract_Two year  \
0              0        1           0                  0                  0   
1              0        0           0                  1                  0   
2              0        0           0                  0                  0   
3              0        0           0                  1                  0   
4              0        0           0                  0                  0   

   ...  OnlineBackup_No  OnlineBackup_Yes  DeviceProtection_No  \
0  ...                0                 1                    1   
1  ...                1                 0                    0   
2  ...                0                 1                    1   
3  ...                1                 0                    0   
4  ...                1                 0                    1   

   DeviceProtection_Yes  TechSupport_No  TechSupport_Yes  StreamingTV_No  \
0                     0               1                0               1   
1                     1               1                0               1   
2                     0               1                0               1   
3                     1               0                1               1   
4                     0               1                0               1   

   StreamingTV_Yes  StreamingMovies_No  StreamingMovies_Yes  
0                0                   1                    0  
1                0                   1                    0  
2                0                   1                    0  
3                0                   1                    0  
4                0                   1                    0  

[5 rows x 30 columns]

In [41]:
# Putting response variable to y
y = telecom['Churn']
y.head()

0    0
1    0
2    1
3    0
4    1
Name: Churn, dtype: int64

In [42]:
# Splitting the data into train and test
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, test_size=0.3, random_state=100)

## Step 5: Feature Scaling

In [43]:
from sklearn.preprocessing import StandardScaler

In [44]:
scaler = StandardScaler()

In [45]:
X_train[['tenure', 'MonthlyCharges', 'TotalCharges']] = scaler.fit_transform(X_train[['tenure', 'MonthlyCharges', 'TotalCharges']])
X_train.head()

tenure  PhoneService  PaperlessBilling  MonthlyCharges  TotalCharges  \
879   0.019693             1                 1       -0.338074     -0.276449   
5790  0.305384             0                 1       -0.464443     -0.112702   
6498 -1.286319             1                 1        0.581425     -0.974430   
880  -0.919003             1                 1        1.505913     -0.550676   
2784 -1.163880             1                 1        1.106854     -0.835971   

      SeniorCitizen  Partner  Dependents  Contract_One year  \
879               0        0           0                  0   
5790              0        1           1                  0   
6498              0        0           0                  0   
880               0        0           0                  0   
2784              0        0           1                  0   

      Contract_Two year  ...  OnlineBackup_No  OnlineBackup_Yes  \
879                   0  ...                0                 1   
5790                  0  ...                0                 1   
6498                  0  ...                0                 1   
880                   0  ...                0                 1   
2784                  0  ...                1                 0   

      DeviceProtection_No  DeviceProtection_Yes  TechSupport_No  \
879                     1                     0               1   
5790                    1                     0               1   
6498                    0                     1               1   
880                     0                     1               0   
2784                    0                     1               0   

      TechSupport_Yes  StreamingTV_No  StreamingTV_Yes  StreamingMovies_No  \
879                 0               1                0                   1   
5790                0               0                1                   0   
6498                0               1                0                   1   
880                 1               0                1                   0   
2784                1               0                1                   0   

      StreamingMovies_Yes  
879                     0  
5790                    1  
6498                    0  
880                     1  
2784                    1  

[5 rows x 30 columns]

In [46]:
### Checking the Churn Rate
churn = (sum(telecom['Churn'])/len(telecom['Churn'].index))*100
churn

26.578498293515356

We have almost 27% churn rate

## Step 6: Looking at Correlations

## Step 7: Model Building

## Step 8: Feature Selection Using RFE

## Step 9: Plotting the ROC Curve

## Step 10: Finding Optimal Cutoff Point

## Step 11: Making predictions on the test set